### PR TITLE
test(g-code-testing): Add module http endpoint tests (RET-152, RET-190)

### PR DESCRIPTION
**Pull request by DerekMaggio**
# Overview

* Added tests for modules
* Refactored test structure
* Added CLI

# Changelog

* Created GCodeConfirmConfig class
  * [ProtocolGCodeConfirm](https://github.com/Opentrons/opentrons/pull/8360/files#diff-0b6de3887342080a5165ed4e1a1601bf0ed2ceff1d3887414f3f26860f526254R83) - Configuration for running protocols. Requires path to protocol
  * [HTTPGCodeConfirmConfig](https://github.com/Opentrons/opentrons/pull/8360/files#diff-0b6de3887342080a5165ed4e1a1601bf0ed2ceff1d3887414f3f26860f526254R98) - Configuration for running HTTP endpoints. Requires executable to run
  * Both classes utilize [SharedFunctionsMixin](https://github.com/Opentrons/opentrons/pull/8360/files#diff-0b6de3887342080a5165ed4e1a1601bf0ed2ceff1d3887414f3f26860f526254R15), which provides functions for:
    * Uploading and downloading files to S3
    * Adding pytest marks to the configuration 
    * Generating pytest parameters to pass to [pytest.mark.parameterize](https://docs.pytest.org/en/6.2.x/parametrize.html#pytest-mark-parametrize-parametrizing-test-functions)
* Added tests for all G-Code producing module endpoints.
  * [magdeck](https://github.com/Opentrons/opentrons/pull/8360/files#diff-7c3fce0439c53c1a05a2d99ddce7fa910ac45712534c8de6d50129e05b1c8178)
  * [tempdeck](https://github.com/Opentrons/opentrons/pull/8360/files#diff-bb406b96ab613118e14b049545a6fd60f7169b1d70da1a22e97cd00d8c0a9788)
  * [thermocycler](https://github.com/Opentrons/opentrons/pull/8360/files#diff-2a3508f10a6b619cdb4bad93b92ca1c6306af4680cc53b2fdaaeb0d7535b47ea)
* Refactored robot control tests in [robot.py](https://github.com/Opentrons/opentrons/pull/8360/files#diff-bd48fca9fcbc3a40e0e0b21c5138cffbbc599f5e820464aeb4cafb2dc50a9f24)
* Composed all HTTP tests together in [http_configurations.py](https://github.com/Opentrons/opentrons/pull/8360/files#diff-d91dd536fe3b53b7d907efc977eef577a5e94035cd2fb6f9f40bb676ef3ba976)
* Defined all protocol tests in [protocol.configurations.py](https://github.com/Opentrons/opentrons/pull/8360/files#diff-bae61fe541c551695453837fb55133ed17922577400d4fc2ece33f4226bf8c0d)
  * Note the marking of, `TWO_MODULES`, `SWIFT_SMOKE`, and `SWIFT_TURBO` as `slow` tests on [this line](https://github.com/Opentrons/opentrons/pull/8360/files#diff-bae61fe541c551695453837fb55133ed17922577400d4fc2ece33f4226bf8c0dR65)
* Combined all HTTP and protocol tests into [test_g_code_comparision.py](https://github.com/Opentrons/opentrons/pull/8360/files#diff-11cda681d9352cb4cfcf164ce8c609d97f4d1f0ad253c195fea52dec36aba8b9)
* Added [cli.py](https://github.com/Opentrons/opentrons/pull/8360/files#diff-6d82c7a7dbfdec60fe5bbe3d098c1ca83bdde563d6a23614bb3b5711b8ff7a10) which supports following commands:
  * `configurations` - Get list of runnable G-Code configurations
  * `run` - Execute G-Code configuration and print output to std out
  * `diff` - Compare execution of G-Code configuration against master file in S3. If files are different print diff
  * `pull` - Pull master file from S3 and print it's content to std out
  * `push` Push execution of G-Code configuration to S3. Will override if it already exists

# Review requests

Is my idea of using the GCodeConfirm classes too weird?

# Risk assessment

Low

### CODE REVIEW

File: g-code-testing/cli.py [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+        """
+        parsed_dict = vars(cls.parser().parse_args(args))
+        return parsed_dict
+
+    @classmethod
+    def _create_configuration_dict(cls) -> Dict[
+        str, Union[HTTPGCodeConfirmConfig, ProtocolGCodeConfirmConfig]
+    ]:
+        """Create lookup dict for configurations"""
+        configurations = {}
```
Comment (amitlissack) [inline g-code-testing/cli.py:48]: Not important to change at all, but unless I'm missing something the body of this func can be replaced with this:
```
configurations = {conf.get_configuration_path(): conf for conf in cls.CONFIGURATIONS}
```

File: g-code-testing/cli.py [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+            text = differ.get_html_diff()
+        else:
+            text = "No difference between compared strings"
+
+            return text
+
+    def _configurations(self) -> str:
+        """Get a list of runnable G-Code Configurations"""
+        path_string = '\n'.join(
+            [
```
Comment (amitlissack) [inline g-code-testing/cli.py:86]: Another unimportant thing. You don't need the `[` `]`. No need to create a list comprehension. A generator comprehension is fine for `join`.

**Review by amitlissack — APPROVED** [review]
Looks good.  Just some minor python comments that don't need to be changed.

If you're using Pycharm you should change the docstring format to `google`.

File: g-code-testing/cli.py (hunk 1 above)
Comment (DerekMaggio) [reply g-code-testing/cli.py:48]: I didn't know this was a thing. In Python 2, dictionary comprehensions don't exist.
I will add this

**Review by sfoster1 — APPROVED** [review]
lgtm!

**Pull request merged**